lib/spack/spack/object_file/elf.py

```diff
@@ -0,0 +1,298 @@
+# Copyright 2013-2021 Lawrence Livermore National Security, LLC and other
+# Spack Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: (Apache-2.0 OR MIT)
+
+import struct
+import sys
+from struct import unpack_from
+
+
+class RPathType(object):
+    #: Don't force an rpath
+    NONE = 0
+    #: Force RPATH
+    RPATH = 1
+    #: Force RUNPATH
+    RUNPATH = 2
+
+
+def get_byte_at(byte_array, idx):
+    if sys.version_info[0] < 3:
+        return ord(byte_array[idx])
+    return byte_array[idx]
+
+
+class ElfParsingError(Exception):
+    pass
+
+
+class ElfFile(object):
+    __slots__ = [
+        'is_64_bit',
+        'is_little_endian',
+        'isa',
+        'dynamic',
+        'rpath_offset',
+        'needed_offsets',
+        'dt_array_rpath_or_runpath_offset',
+        'is_runpath'
+    ]
+
+    def __init__(self, is_64_bit, is_little_endian, isa,
+                 dynamic=False, needed_offsets=None, rpath_offset=None,
+                 dt_array_rpath_or_runpath_offset=None, is_runpath=None):
+        self.is_64_bit = is_64_bit
+        self.is_little_endian = is_little_endian
+        self.isa = isa
+        self.dynamic = dynamic
+        self.rpath_offset = rpath_offset
+        self.needed_offsets = needed_offsets
+        self.dt_array_rpath_or_runpath_offset = dt_array_rpath_or_runpath_offset
+        self.is_runpath = is_runpath
+
+
+def read_c_string_from_file(f, offset, buf_size=1024):
+    # Read C-string till \0 or end of file.
```

Comment:
IMO it's worth adding a `"""`-style docstring and mentioning that your position in the file is "undetermined" after executing the function (e.g. if you read past a `\0`.